src/tiled/scriptpropertytype.h

```diff
@@ -0,0 +1,131 @@
+/*
+ * scriptpropertytype.h
+ * Copyright 2024, chris <dogboydog@users.noreply.github.com>
+ *
+ * This file is part of Tiled.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the Free
+ * Software Foundation; either version 2 of the License, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "propertytype.h"
+
+#include <QJSEngine>
+#include <QList>
+#include <QObject>
+
+namespace Tiled {
+/**
+ * Scripting engine wrapper for PropertyType
+ */
+class ScriptPropertyType : public QObject
+{
+    Q_OBJECT
+    Q_PROPERTY(QString name READ name)
+    Q_PROPERTY(bool isClass READ isClass)
+    Q_PROPERTY(bool isEnum READ isEnum)
+    Q_PROPERTY(QVariant defaultValue READ defaultValue)
+
+public:
+    ScriptPropertyType(const PropertyType *propertyType)
+        : mType(propertyType)
+    {}
+
+    const QString &name() const;
+    bool isClass() const { return mType->isClass(); }
+    bool isEnum() const { return mType->isEnum(); }
+    QVariant defaultValue() { return mType->defaultValue(); }
+
+private:
+    const PropertyType *mType;
+};
+
+class ScriptEnumPropertyType : public ScriptPropertyType
+{
+    Q_OBJECT
+
+    Q_PROPERTY(StorageType storageType READ storageType)
+    Q_PROPERTY(QStringList values READ values)
+
+public:
+    ScriptEnumPropertyType(const EnumPropertyType *propertyType)
+        : ScriptPropertyType(propertyType)
+        , mEnumType(propertyType)
+    {}
+    // copied from propertytype.h
+    enum StorageType {
+        StringValue,
+        IntValue
+    };
+    Q_ENUM(StorageType);
+
+    StorageType storageType() const { return static_cast<StorageType>(mEnumType->storageType); }
+    QStringList values() const { return mEnumType->values; }
+
+private:
+    const EnumPropertyType *mEnumType;
```

Comment:
A raw pointer is problematic here, cause there is nothing that makes sure the `ScriptEnumPropertyType` instance is deleted when the `EnumPropertyType` instance is deleted. There's a few potential solutions we can try:

* Referring to types using shared pointers, so that the wrapper object can keep the type alive.
* Adding a pointer to the `ScriptEnumPropertyType` from the `EnumPropertyType`, so that it can delete the wrapper upon destruction.
* Making a copy of the type that is owned by this wrapper. Though that means changes won't automatically apply, but we'd instead need to have a `Project.addType` or `Project.setType` to assign a modified object back to the project.
* Remembering just the name of the type and looking it up each time it is accessed (might get confusing in case the type is renamed).

I'm not sure at the moment which will be the best way forward. It depends also on the expected usage pattern of the API.

Reply:
To me the first two seem like the best options. I'd prefer to retain the ability to modify a type's fields just by setting them -- I think that's simpler. So that would mean no option 3. Option 4 seems a bit messy for the reasons you mentioned (we may also want to allow renaming the types from scripting I would think, so that would be complicated by option 4)

For option 2, I guess we would have to check if there already has been a ScriptPropertyEnumType created for an EnumPropertyType before we would create a new one. Also if the main types would hold a pointer to the script wrapper,  maybe they should handle creating the script wrapper themselves? 

Option 1 also sounds promising but challenging for my skillset

Reply:
So if we go for the shared pointer route, I wonder where the code for creating the shared pointers would go, maybe in PropertyTypes ?   If we get that working, it would solve the const pointer issue too  and allow the  setters  for color, drawFill, etc. to work